### PR TITLE
feat: Filter on file extensions within search

### PR DESCRIPTION
**Pull request by adrien-coye**
Allows the user to filter files by specific file extension.

IMG_3356

### CODE REVIEW

File: kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift [hunk 1]
```diff
@@ -253,6 +330,10 @@ extension SearchFiltersViewController: SelectDelegate {
         } else if let fileType = option as? ConvertedType {
             filters.fileType = fileType
             reloadSection(.type)
+
+            if fileType == .searchExtension {
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift:334]: Ensure the input cell is selected by default asap

File: kDrive/UI/View/Files/Search/TextInputTableViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        contentInsetView.clipsToBounds = true
+    }
+
+    override func setSelected(_ selected: Bool, animated: Bool) {
+        super.setSelected(selected, animated: animated)
+
+        if selected {
+            contentInsetView.borderWidth = 2
+            contentInsetView.borderColor = KDriveResourcesAsset.infomaniakColor.color
+            contentInsetView.alpha = 1
```
Comment (adrien-coye) [inline kDrive/UI/View/Files/Search/TextInputTableViewCell.swift:41]: TODO clean

File: kDriveCore/Data/Cache/DriveFileManager/DriveFileManager.swift [hunk 3]
```diff
@@ -396,7 +397,14 @@ public final class DriveFileManager {
                                                  drive: drive,
                                                  query: query,
                                                  date: date,
-                                                 fileTypes: [fileType].compactMap { $0 },
+                                                 fileTypes: [fileType].compactMap { type in
+                                                     guard type != .searchExtension else {
```
Comment (adrien-coye) [inline kDriveCore/Data/Cache/DriveFileManager/DriveFileManager.swift:401]: `searchExtension` is not recognized by backed, must be stripped out

File: kDriveCore/Data/Models/File.swift [hunk 4]
```diff
@@ -28,7 +28,8 @@ import QuickLook
 import RealmSwift
 
 public enum ConvertedType: String, CaseIterable {
-    case archive, audio, code, folder = "dir", font, form, image, pdf, presentation, spreadsheet, text, unknown, url, video
+    case archive, audio, code, folder = "dir", font, form, image, pdf, presentation, spreadsheet, text, unknown, url, video,
```
Comment (adrien-coye) [inline kDriveCore/Data/Models/File.swift:31]: I should probably put one by line

File: kDriveCore/Data/Models/File.swift [hunk 5]
```diff
@@ -60,6 +61,8 @@ public enum ConvertedType: String, CaseIterable {
             return KDriveResourcesAsset.url.image
         case .video:
             return KDriveResourcesAsset.fileVideo.image
+        case .searchExtension:
+            return UIImage(systemName: "magnifyingglass")!
```
Comment (adrien-coye) [inline kDriveCore/Data/Models/File.swift:65]: @valentinperignon I used SF Symbols for a quick hack. Works well, will check if there is something homemade to use

File: kDriveCore/Data/Models/File.swift [hunk 6]
```diff
@@ -98,6 +101,9 @@ public enum ConvertedType: String, CaseIterable {
             return ""
         case .video:
             return KDriveResourcesStrings.Localizable.allVideo
+        case .searchExtension:
+            // TODO: i18n
```
Comment (adrien-coye) [inline kDriveCore/Data/Models/File.swift:105]: TODO: i18n

File: kDriveCore/Data/Models/File.swift [hunk 7]
```diff
@@ -131,6 +137,8 @@ public enum ConvertedType: String, CaseIterable {
             return .internetShortcut
         case .video:
             return .movie
+        case .searchExtension:
+            return .item // maybe content ?
```
Comment (adrien-coye) [inline kDriveCore/Data/Models/File.swift:141]: TODO: clean

**Review by adrien-coye — COMMENTED** [review]
Did a self assessment and left some notes for future reviewer

File: kDriveCore/Data/Models/File.swift (hunk 5 above)
Comment (PhilippeWeidmann) [reply kDriveCore/Data/Models/File.swift:65]: We indeed have a search icon: `KDriveResourcesAsset.search.image`

Comment (adrien-coye) [reply kDriveCore/Data/Models/File.swift:65]: ty

File: kDriveCore/Data/Models/File.swift (hunk 4 above)
Comment (adrien-coye) [reply kDriveCore/Data/Models/File.swift:31]: done

File: kDrive/UI/View/Files/Search/TextInputTableViewCell.swift [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import UIKit
+
+final class TextInputTableViewCell: UITableViewCell {
+    @IBOutlet var textField: MaterialOutlinedTextField!
+
+    override func awakeFromNib() {
+        super.awakeFromNib()
+
+        textField.setInfomaniakColors()
+        // TODO: i18n
```
Comment (adrien-coye) [inline kDrive/UI/View/Files/Search/TextInputTableViewCell.swift:33]: i18n

File: kDriveCore/Data/Api/Endpoint.swift [hunk 9]
```diff
@@ -614,6 +615,9 @@ public extension Endpoint {
         for fileType in fileTypes {
             queryItems.append(URLQueryItem(name: "types[]", value: fileType.rawValue))
         }
+        for fileExtension in fileExtensions {
+            queryItems.append(URLQueryItem(name: "extensions[]", value: fileExtension))
```
Comment (adrien-coye) [inline kDriveCore/Data/Api/Endpoint.swift:619]: Matched web API for `extensions[]`

File: kDrive/UI/View/Files/Search/TextInputTableViewCell.swift [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+final class TextInputTableViewCell: UITableViewCell {
+    @IBOutlet var textField: MaterialOutlinedTextField!
+
+    override func awakeFromNib() {
+        super.awakeFromNib()
+
+        textField.setInfomaniakColors()
+        textField.setHint(KDriveResourcesStrings.Localizable.sortExtension)
+        textField.placeholder = ".jpg, .mov …"
```
Comment (valentinperignon) [inline kDrive/UI/View/Files/Search/TextInputTableViewCell.swift:34]:
```suggestion
        textField.placeholder = ".jpg, .mov, etc."
```
To be iso with the webapp 😊

File: kDrive/UI/View/Files/Search/TextInputTableViewCell.swift [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+import Foundation
+import InfomaniakCoreUI
+import kDriveCore
+import kDriveResources
+import MaterialOutlinedTextField
+import UIKit
+
+final class TextInputTableViewCell: UITableViewCell {
```
Comment (valentinperignon) [inline kDrive/UI/View/Files/Search/TextInputTableViewCell.swift:26]: Can we make a more generic view?
And set hint, placeholder and fileExtensionConfiguration when we init the view for the search 👍

File: kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift [hunk 12]
```diff
@@ -222,8 +288,19 @@ class SearchFiltersViewController: UITableViewController {
...
         switch filterType {
-        case .date, .type:
+        case .date:
             break
+        case .type:
+            switch indexPath.row {
+            case 0:
+                break
+            default:
+                // TODO: Matomo?
```
[the remark this answers is not in the document]
Comment (valentinperignon) [reply kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift:302]: Maybe we can add a Matomo?
We can name it "filterExtension", what do you think?

**Review by valentinperignon — COMMENTED** [review]
When we come back on the filters view, the textField is empty

--
Great work! 👍

File: kDrive/UI/View/Files/Search/TextInputTableViewCell.swift (hunk 10 above)
Comment (adrien-coye) [reply kDrive/UI/View/Files/Search/TextInputTableViewCell.swift:34]: discussed, I keep it this way.

File: kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift:302]: Perfect 👍

File: kDrive/UI/View/Files/Search/TextInputTableViewCell.swift (hunk 11 above)
Comment (adrien-coye) [reply kDrive/UI/View/Files/Search/TextInputTableViewCell.swift:26]: Great idea 👍

File: kDriveCore/Data/Api/Endpoint.swift (hunk 9 above)
Comment (adrien-coye) [reply kDriveCore/Data/Api/Endpoint.swift:619]: Waiting for `search/default` to be deprecated and migrated into `search`

**Comment by adrien-coye** [comment]
Remade the cell to be code based to ensure easy to use inheritance.
Also, reopening the filter selection will persist extensions.
Just

File: kDriveCore/Data/Api/Endpoint.swift (hunk 9 above)
Comment (adrien-coye) [reply kDriveCore/Data/Api/Endpoint.swift:619]: API in production. Checked that it works. All good. Merging.

File: kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift [hunk 13]
```diff
@@ -64,6 +71,7 @@ class SearchFiltersViewController: UITableViewController {
         tableView.register(cellView: LocationTableViewCell.self)
         tableView.register(cellView: ManageCategoriesTableViewCell.self)
         tableView.register(cellView: SelectTableViewCell.self)
+        tableView.register(FileExtensionTextInputTableViewCell.self, forCellReuseIdentifier: "FileExtensionTextInputTableViewCell")
```
Comment (valentinperignon) [inline kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift:74]: I think we should use `tableView.register(cellView:)`

File: kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift [hunk 14]
```diff
@@ -99,6 +107,41 @@ class SearchFiltersViewController: UITableViewController {
         }
     }
 
+    // MARK: - TextFieldDelegate
```
Comment (valentinperignon) [inline kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift:110]: Maybe we can move the delegate conformance in an extension?

File: kDrive/UI/View/Files/Search/FileExtensionTextInputTableViewCell.swift [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+import kDriveResources
+import MaterialOutlinedTextField
+import UIKit
+
+final class FileExtensionTextInputTableViewCell: TextInputTableViewCell {
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+
+        textField.setHint(KDriveResourcesStrings.Localizable.sortExtension)
+        textField.placeholder = ".jpg, .mov …"
```
Comment (valentinperignon) [inline kDrive/UI/View/Files/Search/FileExtensionTextInputTableViewCell.swift:31]:
```suggestion
        textField.placeholder = ".jpg, .mov, etc."
```

Comment (adrien-coye) [reply kDrive/UI/View/Files/Search/FileExtensionTextInputTableViewCell.swift:31]: We did discuss that previously, if I use `etc`it should be internationalized.

File: kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift (hunk 14 above)
Comment (adrien-coye) [reply kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift:110]: sure.

File: kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift (hunk 13 above)
Comment (adrien-coye) [reply kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift:74]: I had issues with it during the rewrite in pure code of the cell. It does not work if there is no XIB.
I did it again just to be sure, and I had this exception : `"Could not load NIB in bundle: `

Comment (valentinperignon) [reply kDrive/UI/Controller/Files/Search/SearchFiltersViewController.swift]: Oh right, the method create a UINib... Thank you for trying again, I resolve the conversation.

File: kDrive/UI/View/Files/Search/FileExtensionTextInputTableViewCell.swift (hunk 15 above)
Comment (valentinperignon) [reply kDrive/UI/View/Files/Search/FileExtensionTextInputTableViewCell.swift:31]: Yeah, I think it should be localized! We should use the same wording as the web app.

**Review by valentinperignon — APPROVED** [review]

**Pull request merged**